sui_types/src/waypoint.rs

```diff
@@ -0,0 +1,457 @@
+    {
+        for i in items {
+            self.insert(i);
+        }
+    }
+}
+
+impl std::fmt::Debug for Accumulator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Accumulator").finish()
+    }
+}
+
+/*
+    A way point represents a sequential point in a stream that summarizes
+    all elements so far. Waypoints with increasing sequence numbers should
+    contain all element in previous waypoints and expand them.
+*/
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct Waypoint {
+    pub sequence_number: u64,
+    pub accumulator: Accumulator,
+}
+
+impl Waypoint {
+    /// Make a new waypoint.
+    pub fn new(sequence_number: u64) -> Waypoint {
+        Waypoint {
+            sequence_number,
+            accumulator: Accumulator::default(),
+        }
+    }
+
+    /// Inserts an element into the accumulator
+    pub fn insert<I>(&mut self, item: &I)
+    where
+        I: Borrow<[u8]>,
+    {
+        self.accumulator.insert(item);
+    }
+}
+
+/*
+    A structure to hold a waypoint, associated items,
+    and is indexed by a key provided. Such a structure
+    may be used to represent checkpoints or differences
+    of checkpoints.
+*/
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct WaypointWithItems<K, I>
+where
+    K: Clone,
+    I: Ord,
+{
+    pub key: K,
+    pub waypoint: Waypoint,
+    pub items: BTreeSet<I>,
+}
+
+impl<K, I> WaypointWithItems<K, I>
+where
+    K: Clone,
+    I: Borrow<[u8]> + Ord,
+{
+    pub fn new(key: K, sequence_number: u64) -> WaypointWithItems<K, I> {
+        WaypointWithItems {
+            key,
+            waypoint: Waypoint::new(sequence_number),
+            items: BTreeSet::new(),
+        }
+    }
+
+    /// Insert an element in the accumulator and list of items
+    pub fn insert_full(&mut self, item: I) {
+        self.waypoint.accumulator.insert(&item);
+        self.items.insert(item);
+    }
+
+    /// Insert an element in the accumulator only
+    pub fn insert_accumulator(&mut self, item: &I) {
+        self.waypoint.accumulator.insert(item);
+    }
+
+    /// Insert an element in the items only
+    pub fn insert_item(&mut self, item: I) {
+        self.items.insert(item);
+    }
+}
+
+/*
+    Represents the difference between two waypoints
+    and elements that make up this difference.
+*/
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct WaypointDiff<K, I>
+where
+    K: Clone,
+    I: Borrow<[u8]> + Ord,
+{
+    pub first: WaypointWithItems<K, I>,
+    pub second: WaypointWithItems<K, I>,
+}
+
+impl<K, I> WaypointDiff<K, I>
+where
+    K: Clone,
+    I: Borrow<[u8]> + Ord,
+{
+    pub fn new<V1, V2>(
+        first_key: K,
+        first: Waypoint,
+        missing_from_first: V1,
+        second_key: K,
+        second: Waypoint,
+        missing_from_second: V2,
+    ) -> WaypointDiff<K, I>
+    where
+        V1: Iterator<Item = I>,
+        V2: Iterator<Item = I>,
```

Comment:
I needed to have distinct types for V1 and V2 to allow the use of different combinators when passing in iterators for the two sides. Since ultimately V1 and V2 will be instantiated with the same type, and the result of combinators on streams leads to different types, having both of them as V forces a unification that is awkward. 

I have now changed the trait to IntoItererator! 

In term of making `I` more flexible, we already have the bound on the `impl` block `I: AsRef<[u8]> + Ord`. I think this will also cover this function delivering the functionality / flexibility you suggest.